### ISSUE TITLE
forkserver could warn if several threads are running

### ISSUE THREAD

**Issue by pitrou** [python/cpython#76094]
BPO | [31913](https://bugs.python.org/issue31913)
--- | :---
Nosy | @gpshead, @pitrou, @giampaolo, @applio

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-10-31.18:12:26.725>
labels = ['3.7', 'type-feature', 'library']
title = 'forkserver could warn if several threads are running'
updated_at = <Date 2017-11-01.23:22:22.238>
user = 'https://github.com/pitrou'
```

bugs.python.org fields:
```python
activity = <Date 2017-11-01.23:22:22.238>
actor = 'pitrou'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-10-31.18:12:26.725>
creator = 'pitrou'
dependencies = []
files = []
hgrepos = []
issue_num = 31913
keywords = []
message_count = 3.0
messages = ['305323', '305400', '305401']
nosy_count = 4.0
nosy_names = ['gregory.p.smith', 'pitrou', 'giampaolo.rodola', 'davin']
pr_nums = []
priority = 'low'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue31913'
versions = ['Python 3.7']
```

</p></details>

**Comment by pitrou** [comment]
I'm not sure this is worth handling, but I had an interaction with a user who had weird deadlock problems in a glibc function (getaddrinfo) in worker processes launched with the forkserver method.

The explanation turned out to be that a sitecustomize.py did some stuff that eventually launched a helper thread, and that made the forkserver process's forking fundamentally unsafe (fork() is guaranteed to be safe if there's only one thread running in the parent process).

It would be easy to check that threading.enumerate() returns only one thread, and otherwise warn the user about it.  Note this only handles Python threads and not C threads invisible to Python... (a more complete solution would involve psutil :-)).

**Comment by gpshead** [comment]
If we do this, I think we should aim for the complete solution on the most common posix platforms (Linux and MacOS) as C/C++ extensions are just as happy to create threads these days.  (but if you want to start with a simple python threads only warning, i'm not going to stop you)

**Comment by pitrou** [comment]
Yeah... We might replicate the psutil source code doing that.  On Linux, it involves parsing /proc/{pid}/status (which may not be available on some restricted execution environments?).  Haven't looked what it does on macOS.
